In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import train_test_split
pd.set_option("display.max_rows",None)
pd.set_option("display.max_columns",None)

In [2]:
ruta="./Datos/"
train=os.listdir(ruta)[1]
df=pd.read_csv(ruta+train)

In [3]:
df.head()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               Y             0           67500.0    135000.0       6750.0   
3               Y             0          135000.0    312682.5      29686.5   
4               Y             0          121500.0    513000.0      21865.5   

   AMT_GOODS_PRICE NAME_TYPE_SUITE NAME_INCOME_TYPE  \
0         351000.0   Unaccompanied          Working   
1        1129500.0          Family    State servant   
2         135000.0   Unaccompanied          Working   
3         297000.0   Unaccompanied          Working   
4         513000.0   Unaccompanied          Working   

             NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  \
0  Secondary / secondary special  Single / not married  House / apartment   
1               Higher education               Married  House / apartment   
2  Secondary / secondary special  Single / not married  House / apartment   
3  Secondary / secondary special        Civil marriage  House / apartment   
4  Secondary / secondary special  Single / not married  House / apartment   

   REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  \
0                    0.018801       -9461           -637            -3648.0   
1                    0.003541      -16765          -1188            -1186.0   
2                    0.010032      -19046           -225            -4260.0   
3                    0.008019      -19005          -3039            -9833.0   
4                    0.028663      -19932          -3038            -4311.0   

   DAYS_ID_PUBLISH  OWN_CAR_AGE  FLAG_MOBIL  FLAG_EMP_PHONE  FLAG_WORK_PHONE  \
0            -2120          NaN           1               1                0   
1             -291          NaN           1               1                0   
2            -2531         26.0           1               1                1   
3            -2437          NaN           1               1                0   
4            -3458          NaN           1               1                0   

   FLAG_CONT_MOBILE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  CNT_FAM_MEMBERS  \
0                 1           1           0        Laborers              1.0   
1                 1           1           0      Core staff              2.0   
2                 1           1           0        Laborers              1.0   
3                 1           0           0        Laborers              2.0   
4                 1           0           0      Core staff              1.0   

   REGION_RATING_CLIENT  REGION_RATING_CLIENT_W_CITY  \
0                     2                            2   
1                     1                            1   
2                     2                            2   
3                     2                            2   
4                     2                            2   

  WEEKDAY_APPR_PROCESS_START  HOUR_APPR_PROCESS_START  \
0                  WEDNESDAY                       10   
1                     MONDAY                       11   
2                     MONDAY                        9   
3                  WEDNESDAY                       17   
4                   THURSDAY                       11   

   REG_REGION_NOT_LIVE_REGION  REG_REGION_NOT_WORK_REGION  \
0                           0                           0   
1                           0                           0   
2                        

In [4]:
df.describe(include="all")

SK_ID_CURR         TARGET NAME_CONTRACT_TYPE CODE_GENDER  \
count   307511.000000  307511.000000             307511      307511   
unique            NaN            NaN                  2           3   
top               NaN            NaN         Cash loans           F   
freq              NaN            NaN             278232      202448   
mean    278180.518577       0.080729                NaN         NaN   
std     102790.175348       0.272419                NaN         NaN   
min     100002.000000       0.000000                NaN         NaN   
25%     189145.500000       0.000000                NaN         NaN   
50%     278202.000000       0.000000                NaN         NaN   
75%     367142.500000       0.000000                NaN         NaN   
max     456255.000000       1.000000                NaN         NaN   

       FLAG_OWN_CAR FLAG_OWN_REALTY   CNT_CHILDREN  AMT_INCOME_TOTAL  \
count        307511          307511  307511.000000      3.075110e+05   
unique            2               2            NaN               NaN   
top               N               Y            NaN               NaN   
freq         202924          213312            NaN               NaN   
mean            NaN             NaN       0.417052      1.687979e+05   
std             NaN             NaN       0.722121      2.371231e+05   
min             NaN             NaN       0.000000      2.565000e+04   
25%             NaN             NaN       0.000000      1.125000e+05   
50%             NaN             NaN       0.000000      1.471500e+05   
75%             NaN             NaN       1.000000      2.025000e+05   
max             NaN             NaN      19.000000      1.170000e+08   

          AMT_CREDIT    AMT_ANNUITY  AMT_GOODS_PRICE NAME_TYPE_SUITE  \
count   3.075110e+05  307499.000000     3.072330e+05          306219   
unique           NaN            NaN              NaN               7   
top              NaN            NaN              NaN   Unaccompanied   
freq             NaN            NaN              NaN          248526   
mean    5.990260e+05   27108.573909     5.383962e+05             NaN   
std     4.024908e+05   14493.737315     3.694465e+05             NaN   
min     4.500000e+04    1615.500000     4.050000e+04             NaN   
25%     2.700000e+05   16524.000000     2.385000e+05             NaN   
50%     5.135310e+05   24903.000000     4.500000e+05             NaN   
75%     8.086500e+05   34596.000000     6.795000e+05             NaN   
max     4.050000e+06  258025.500000     4.050000e+06             NaN   

       NAME_INCOME_TYPE            NAME_EDUCATION_TYPE NAME_FAMILY_STATUS  \
count            307511                         307511             307511   
unique                8                              5                  6   
top             Working  Secondary / secondary special            Married   
freq             158774                         218391             196432   
mean                NaN                            NaN                NaN   
std                 NaN                            NaN                NaN   
min                 NaN                            NaN                NaN   
25%                 NaN                            NaN                NaN   
50%                 NaN                            NaN                NaN   
75%                 NaN                            NaN                NaN   
max                 NaN                            NaN                NaN   

        NAME_HOUSING_TYPE  REGION_POPULATION_RELATIVE     DAYS_BIRTH  \
count              307511               307511.000000  307511.000000   
unique                  6                         NaN            NaN   
top     House / apartment                         NaN            NaN   
freq               272868                         NaN            NaN   
mean                  NaN                    0.020868  -16036.995067   
std                   NaN                    0.013831    4363.988632   

In [5]:
import seaborn as sns
Flag=[x for x in df.columns if "FLAG" in x]

In [6]:
# plt.figure(figsize=(15,20))
# cor_matrix=df[Flag].corr()
# sns.heatmap(cor_matrix,annot=True)

In [7]:
No_Flag=[x for x in df.columns if "FLAG" not in x]

In [8]:
# cor_matrix=df[No_Flag].select_dtypes(include="number").corr()
# sns.heatmap(cor_matrix,annot=True)

Pasos a seguir:
* Calcularemos los diferentes valores que tienen las variables objeto.
* Las variables objeto utilizaremos un getdummies escogiedno solo el top5.
* Los valores nulos los imputaremos mediante arbol de decisión haciendo antes un KMEANS.
* Tendremos que balancear la target si queremos entrenar ya que el 8% es uno.
* Una vez imputados haremos cluster para predecir.
* Utilizaremos el arbol de decisión haciendo un gridsearchv.
* Utilizaremos application_Test para predecir.

In [9]:
for i in df.select_dtypes(include="object").columns:
    print(f"La variable {i} tiene estos valores únicos {df[i].nunique()}")

La variable NAME_CONTRACT_TYPE tiene estos valores únicos 2
La variable CODE_GENDER tiene estos valores únicos 3
La variable FLAG_OWN_CAR tiene estos valores únicos 2
La variable FLAG_OWN_REALTY tiene estos valores únicos 2
La variable NAME_TYPE_SUITE tiene estos valores únicos 7
La variable NAME_INCOME_TYPE tiene estos valores únicos 8
La variable NAME_EDUCATION_TYPE tiene estos valores únicos 5
La variable NAME_FAMILY_STATUS tiene estos valores únicos 6
La variable NAME_HOUSING_TYPE tiene estos valores únicos 6
La variable OCCUPATION_TYPE tiene estos valores únicos 18
La variable WEEKDAY_APPR_PROCESS_START tiene estos valores únicos 7
La variable ORGANIZATION_TYPE tiene estos valores únicos 58
La variable FONDKAPREMONT_MODE tiene estos valores únicos 4
La variable HOUSETYPE_MODE tiene estos valores únicos 3
La variable WALLSMATERIAL_MODE tiene estos valores únicos 7
La variable EMERGENCYSTATE_MODE tiene estos valores únicos 2


Observamos la variable que más diferentes valores unicos tiene

In [10]:
sorted(df["ORGANIZATION_TYPE"].unique())

['Advertising',
 'Agriculture',
 'Bank',
 'Business Entity Type 1',
 'Business Entity Type 2',
 'Business Entity Type 3',
 'Cleaning',
 'Construction',
 'Culture',
 'Electricity',
 'Emergency',
 'Government',
 'Hotel',
 'Housing',
 'Industry: type 1',
 'Industry: type 10',
 'Industry: type 11',
 'Industry: type 12',
 'Industry: type 13',
 'Industry: type 2',
 'Industry: type 3',
 'Industry: type 4',
 'Industry: type 5',
 'Industry: type 6',
 'Industry: type 7',
 'Industry: type 8',
 'Industry: type 9',
 'Insurance',
 'Kindergarten',
 'Legal Services',
 'Medicine',
 'Military',
 'Mobile',
 'Other',
 'Police',
 'Postal',
 'Realtor',
 'Religion',
 'Restaurant',
 'School',
 'Security',
 'Security Ministries',
 'Self-employed',
 'Services',
 'Telecom',
 'Trade: type 1',
 'Trade: type 2',
 'Trade: type 3',
 'Trade: type 4',
 'Trade: type 5',
 'Trade: type 6',
 'Trade: type 7',
 'Transport: type 1',
 'Transport: type 2',
 'Transport: type 3',
 'Transport: type 4',
 'University',
 'XNA']

Extraeremos hasta el espacio de cada una y eliminamos cosas que sobren

In [11]:
df["ORGANIZATION_TYPE"]=df["ORGANIZATION_TYPE"].str.split().str[0].str.replace(":", "")

## Pareto
* Se Creará un diccionario para aplicar la regla de pareto

In [12]:
for col in df.select_dtypes(include="object").columns:
    print("="*20)
    print(f"La columna es {col}")
    display(pd.DataFrame(df[col].value_counts(normalize=True).cumsum()))

La columna es NAME_CONTRACT_TYPE


NAME_CONTRACT_TYPE
Cash loans                 0.904787
Revolving loans            1.000000

La columna es CODE_GENDER


CODE_GENDER
F       0.658344
M       0.999987
XNA     1.000000

La columna es FLAG_OWN_CAR


FLAG_OWN_CAR
N      0.659892
Y      1.000000

La columna es FLAG_OWN_REALTY


FLAG_OWN_REALTY
Y         0.693673
N         1.000000

La columna es NAME_TYPE_SUITE


NAME_TYPE_SUITE
Unaccompanied           0.811596
Family                  0.942708
Spouse, partner         0.979838
Children                0.990507
Other_B                 0.996287
Other_A                 0.999115
Group of people         1.000000

La columna es NAME_INCOME_TYPE


NAME_INCOME_TYPE
Working                       0.516320
Commercial associate          0.749212
Pensioner                     0.929245
State servant                 0.999821
Unemployed                    0.999893
Student                       0.999951
Businessman                   0.999984
Maternity leave               1.000000

La columna es NAME_EDUCATION_TYPE


NAME_EDUCATION_TYPE
Secondary / secondary special             0.710189
Higher education                          0.953637
Incomplete higher                         0.987057
Lower secondary                           0.999467
Academic degree                           1.000000

La columna es NAME_FAMILY_STATUS


NAME_FAMILY_STATUS
Married                         0.638780
Single / not married            0.786560
Civil marriage                  0.883386
Separated                       0.947677
Widow                           0.999993
Unknown                         1.000000

La columna es NAME_HOUSING_TYPE


NAME_HOUSING_TYPE
House / apartment             0.887344
With parents                  0.935602
Municipal apartment           0.971968
Rented apartment              0.987841
Office apartment              0.996351
Co-op apartment               1.000000

La columna es OCCUPATION_TYPE


OCCUPATION_TYPE
Laborers                      0.261396
Sales staff                   0.413452
Core staff                    0.544041
Managers                      0.645268
Drivers                       0.733384
High skill tech staff         0.787287
Accountants                   0.833768
Medicine staff                0.874204
Security staff                0.906039
Cooking staff                 0.934203
Cleaning staff                0.956243
Private service staff         0.968804
Low-skill Laborers            0.978718
Waiters/barmen staff          0.985103
Secretaries                   0.991285
Realty agents                 0.994842
HR staff                      0.997509
IT staff                      1.000000

La columna es WEEKDAY_APPR_PROCESS_START


WEEKDAY_APPR_PROCESS_START
TUESDAY                      0.175282
WEDNESDAY                    0.344167
MONDAY                       0.509084
THURSDAY                     0.673602
FRIDAY                       0.837297
SATURDAY                     0.947381
SUNDAY                       1.000000

La columna es ORGANIZATION_TYPE


ORGANIZATION_TYPE
Business                0.274881
XNA                     0.454953
Self-employed           0.579865
Other                   0.634117
Trade                   0.680668
Industry                0.727207
Medicine                0.763605
Government              0.797438
Transport               0.826673
School                  0.855592
Kindergarten            0.877965
Construction            0.899821
Security                0.916800
Housing                 0.926419
Military                0.934984
Bank                    0.943137
Agriculture             0.951117
Police                  0.958730
Postal                  0.965744
Restaurant              0.971634
Services                0.976755
University              0.981071
Hotel                   0.984212
Electricity             0.987301
Insurance               0.989243
Telecom                 0.991119
Emergency               0.992940
Advertising             0.994335
Realtor                 0.995623
Culture                 0.996855
Mobile                  0.997886
Legal                   0.998878
Cleaning                0.999724
Religion                1.000000

La columna es FONDKAPREMONT_MODE


FONDKAPREMONT_MODE
reg oper account                 0.759443
reg oper spec account            0.883702
not specified                    0.942201
org spec account                 1.000000

La columna es HOUSETYPE_MODE


HOUSETYPE_MODE
block of flats          0.982306
specific housing        0.992089
terraced house          1.000000

La columna es WALLSMATERIAL_MODE


WALLSMATERIAL_MODE
Panel                   0.436859
Stone, brick            0.865615
Block                   0.926824
Wooden                  0.962294
Mixed                   0.977482
Monolithic              0.989251
Others                  1.000000

La columna es EMERGENCYSTATE_MODE


EMERGENCYSTATE_MODE
No              0.985608
Yes             1.000000

* En este ejercicio se pondrá un umbral del 90% y se sumará una posición por si acaso se le sumará una posición
* Se aplciará a aquellas  variables uqe superen las siete categoriás diferentes

In [13]:
Diccionario={}
# col="ORGANIZATION_TYPE"
for col in df.select_dtypes(include="object").columns:
    if df[col].nunique()>7:
        ss=pd.DataFrame(df[col].value_counts(normalize=True).cumsum())
        Diccionario[col]=list(ss[:len(ss[ss[col]<=0.9])+1].index)

In [14]:
Diccionario

{'NAME_INCOME_TYPE': ['Working', 'Commercial associate', 'Pensioner'],
 'OCCUPATION_TYPE': ['Laborers',
  'Sales staff',
  'Core staff',
  'Managers',
  'Drivers',
  'High skill tech staff',
  'Accountants',
  'Medicine staff',
  'Security staff'],
 'ORGANIZATION_TYPE': ['Business',
  'XNA',
  'Self-employed',
  'Other',
  'Trade',
  'Industry',
  'Medicine',
  'Government',
  'Transport',
  'School',
  'Kindergarten',
  'Construction',
  'Security']}

* Se detecta que XNA son valores nulos por lo que se deben sustuir

In [15]:
df=df.replace('XNA',np.nan)

In [16]:
df.isnull().sum()

SK_ID_CURR                           0
TARGET                               0
NAME_CONTRACT_TYPE                   0
CODE_GENDER                          4
FLAG_OWN_CAR                         0
FLAG_OWN_REALTY                      0
CNT_CHILDREN                         0
AMT_INCOME_TOTAL                     0
AMT_CREDIT                           0
AMT_ANNUITY                         12
AMT_GOODS_PRICE                    278
NAME_TYPE_SUITE                   1292
NAME_INCOME_TYPE                     0
NAME_EDUCATION_TYPE                  0
NAME_FAMILY_STATUS                   0
NAME_HOUSING_TYPE                    0
REGION_POPULATION_RELATIVE           0
DAYS_BIRTH                           0
DAYS_EMPLOYED                        0
DAYS_REGISTRATION                    0
DAYS_ID_PUBLISH                      0
OWN_CAR_AGE                     202929
FLAG_MOBIL                           0
FLAG_EMP_PHONE                       0
FLAG_WORK_PHONE                      0
FLAG_CONT_MOBILE         

Se ve variable como ``OWN_CAR_AGE`` que se deberá estudiar pq tiene 202929 valores nuloes.

In [17]:
df["OWN_CAR_AGE"].describe()

count    104582.000000
mean         12.061091
std          11.944812
min           0.000000
25%           5.000000
50%           9.000000
75%          15.000000
max          91.000000
Name: OWN_CAR_AGE, dtype: float64

Esta variable puede ser muy controvertida ya que nos falta información.
Por ello crearemos una variable de control de si tiene nan le daremos un 1 y sino un cero y lo compararemos con la target

In [18]:
df["Flag_OWN_CAR_AGE"]=np.where(df["OWN_CAR_AGE"].isnull(),1,0)

In [19]:
pd.crosstab(df["Flag_OWN_CAR_AGE"],df["TARGET"],margins=True,normalize=True)

TARGET                   0         1       All
Flag_OWN_CAR_AGE                              
0                 0.315455  0.024637  0.340092
1                 0.603816  0.056092  0.659908
All               0.919271  0.080729  1.000000

El 5% de las personas que no tienen coche propio tienen un target como uno.
Al estar desbalaceado la muestra a la edad del coche se le pondrá un valor alto ``999`` por si creamos un KMEANS y un modelo de ML Supervisado lo detecte en otro grupo

In [20]:
df["OWN_CAR_AGE"].fillna(999,inplace=True)

In [21]:
df.isnull().sum().sort_values(ascending=False)


COMMONAREA_MODE                 214865
COMMONAREA_AVG                  214865
COMMONAREA_MEDI                 214865
NONLIVINGAPARTMENTS_MEDI        213514
NONLIVINGAPARTMENTS_MODE        213514
NONLIVINGAPARTMENTS_AVG         213514
FONDKAPREMONT_MODE              210295
LIVINGAPARTMENTS_AVG            210199
LIVINGAPARTMENTS_MODE           210199
LIVINGAPARTMENTS_MEDI           210199
FLOORSMIN_AVG                   208642
FLOORSMIN_MODE                  208642
FLOORSMIN_MEDI                  208642
YEARS_BUILD_MEDI                204488
YEARS_BUILD_AVG                 204488
YEARS_BUILD_MODE                204488
LANDAREA_AVG                    182590
LANDAREA_MODE                   182590
LANDAREA_MEDI                   182590
BASEMENTAREA_MODE               179943
BASEMENTAREA_MEDI               179943
BASEMENTAREA_AVG                179943
EXT_SOURCE_1                    173378
NONLIVINGAREA_AVG               169682
NONLIVINGAREA_MODE              169682
NONLIVINGAREA_MEDI       

Para hacer diferentes imputaciones haremos lo siguiente.
* Los registros que tengan menos del 5% de valores nulos lo imputaremos a la moda **por si acaso es una variable categorica** o mediana por si fuera una variable ordinal
* Los registros que tengan más del 5% y menos del 40% de valores nulos lo imputaremos mediante un arbol de decisión.
* Los registros que superen el 40% e inferior al 60% haremos KMEANS entre 2 y 4 cluster y un arbol de decisión.
* Los registros que superen el 60% serán eliminados

In [22]:
Imputaciones=pd.DataFrame(df.isnull().sum().sort_values(ascending=False)/len(df)).reset_index()
Imputaciones.columns=["Columnas","Porcentaje"]

In [23]:
Eliminar=Imputaciones[Imputaciones["Porcentaje"]>=0.5]["Columnas"].values
Mediana=Imputaciones[Imputaciones["Porcentaje"]<0.05]["Columnas"].values
Variables_Arboles=Imputaciones[(Imputaciones["Porcentaje"]>=0.05) & (Imputaciones["Porcentaje"]<0.4)]["Columnas"].values
Variables_Karbol=Imputaciones[(Imputaciones["Porcentaje"]>=0.4) & (Imputaciones["Porcentaje"]<0.5)]["Columnas"].values

### Eliminadas

In [24]:
df=df.drop(columns=Eliminar)

### Imputadas a la mediana

In [25]:
import scipy.stats as ss

In [26]:
for col in Mediana:
    if df[col].dtypes=='O':
        relleno=df[col].value_counts().index[0]
        df[col]=df[col].fillna(relleno)
    elif df[col].dtypes=='int64':
         df[col]=df[col].fillna(np.median(df[col]))
    else:
        df[col]=df[col].fillna((df[col].mean()))

In [27]:
pd.DataFrame(df.isnull().sum().sort_values(ascending=False)/len(df)).reset_index()

index         0
0                 FLOORSMAX_MEDI  0.497608
1                 FLOORSMAX_MODE  0.497608
2                  FLOORSMAX_AVG  0.497608
3   YEARS_BEGINEXPLUATATION_MEDI  0.487810
4   YEARS_BEGINEXPLUATATION_MODE  0.487810
5    YEARS_BEGINEXPLUATATION_AVG  0.487810
6                 TOTALAREA_MODE  0.482685
7            EMERGENCYSTATE_MODE  0.473983
8                OCCUPATION_TYPE  0.313455
9                   EXT_SOURCE_3  0.198253
10             ORGANIZATION_TYPE  0.180072
11    AMT_REQ_CREDIT_BUREAU_HOUR  0.135016
12     AMT_REQ_CREDIT_BUREAU_DAY  0.135016
13    AMT_REQ_CREDIT_BUREAU_WEEK  0.135016
14     AMT_REQ_CREDIT_BUREAU_MON  0.135016
15     AMT_REQ_CREDIT_BUREAU_QRT  0.135016
16    AMT_REQ_CREDIT_BUREAU_YEAR  0.135016
17               FLAG_DOCUMENT_4  0.000000
18      OBS_30_CNT_SOCIAL_CIRCLE  0.000000
19      DEF_30_CNT_SOCIAL_CIRCLE  0.000000
20      OBS_60_CNT_SOCIAL_CIRCLE  0.000000
21      DEF_60_CNT_SOCIAL_CIRCLE  0.000000
22        DAYS_LAST_PHONE_CHANGE  0.000000
23               FLAG_DOCUMENT_2  0.000000
24               FLAG_DOCUMENT_3  0.000000
25                    SK_ID_CURR  0.000000
26               FLAG_DOCUMENT_5  0.000000
27              FLAG_DOCUMENT_15  0.000000
28              FLAG_DOCUMENT_21  0.000000
29              FLAG_DOCUMENT_20  0.000000
30              FLAG_DOCUMENT_19  0.000000
31              FLAG_DOCUMENT_18  0.000000
32              FLAG_DOCUMENT_17  0.000000
33              FLAG_DOCUMENT_16  0.000000
34              FLAG_DOCUMENT_14  0.000000
35               FLAG_DOCUMENT_6  0.000000
36              FLAG_DOCUMENT_13  0.000000
37              FLAG_DOCUMENT_12  0.000000
38              FLAG_DOCUMENT_11  0.000000
39              FLAG_DOCUMENT_10  0.000000
40               FLAG_DOCUMENT_9  0.000000
41               FLAG_DOCUMENT_7  0.000000
42               FLAG_DOCUMENT_8  0.000000
43                  EXT_SOURCE_2  0.000000
44                        TARGET  0.000000
45               AMT_GOODS_PRICE  0.000000
46                    DAYS_BIRTH  0.000000
47    REGION_POPULATION_RELATIVE  0.000000
48             NAME_HOUSING_TYPE  0.000000
49            NAME_FAMILY_STATUS  0.000000
50           NAME_EDUCATION_TYPE  0.000000
51              NAME_INCOME_TYPE  0.000000
52               NAME_TYPE_SUITE  0.000000
53                   AMT_ANNUITY  0.000000
54             DAYS_REGISTRATION  0.000000
55                    AMT_CREDIT  0.000000
56              AMT_INCOME_TOTAL  0.000000
57                  CNT_CHILDREN  0.000000
58               FLAG_OWN_REALTY  0.000000
59                  FLAG_OWN_CAR  0.000000
60                   CODE_GENDER  0.000000
61            NAME_CONTRACT_TYPE  0.000000
62                 DAYS_EMPLOYED  0.000000
63               DAYS_ID_PUBLISH  0.000000
64       LIVE_CITY_NOT_WORK_CITY  0.000000
65   REGION_RATING_CLIENT_W_CITY  0.000000
66        REG_CITY_NOT_WORK_CITY  0.000000
67        REG_CITY_NOT_LIVE_CITY  0.000000
68   LIVE_REGION_NOT_WORK_REGION  0.000000
69    REG_REGION_NOT_WORK_REGION  0.000000
70    REG_REGION_NOT_LIVE_REGION  0.000000
71       HOUR_APPR_PROCESS_START  0.000000
72    WEEKDAY_APPR_PROCESS_START  0.000000
73          REGION_RATING_CLIENT  0.000000
74                   OWN_CAR_AGE  0.000000
75               CNT_FAM_MEMBERS  0.000000
76                    FLAG_EMAIL  0.000000
77                    FLAG_PHONE  0.000000
78              FLAG_CONT_MOBILE  0.000000
79               FLAG_WORK_PHONE  0.000000
80                FLAG_EMP_PHONE  0.000000
81                    FLAG_MOBIL  0.000000
82              Flag_OWN_CAR_AGE  0.000000

## Arbol de Decisión

Ahora viene el paso de hacer arboles de decisión
* Los que sean valores enteros lo consideraremos de clasificación
* los valores flotantes como de regresión

In [28]:
from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor
from sklearn.preprocessing import LabelEncoder

Dos opciones que tenemos aqui como en el siguiente apartado:
1. Que las que se han entrenado y predicho entren para predecir la siguiente
2. Que se entrenen solo con las que no tienen nulos hasta este punto y se predice cada una.

```para esta solución optamos por la segunda```

Además aquellas variables que sean objeto no la utilizaremos para entrenar, para así evitarnos utilziar el labeleconder y asi evitar si una variable no ha sido vista por el label enconer no tenga problemas

In [29]:
Variables_Entrenamiento_Posibles=pd.DataFrame(df.isnull().sum().sort_values(ascending=False)/len(df)).reset_index()
Variables_Entrenamiento_Posibles.columns=["Columnas","Porcentaje"]
Variables_Entrenamiento_Posibles=Variables_Entrenamiento_Posibles[Variables_Entrenamiento_Posibles["Porcentaje"]==0.0]["Columnas"].values

In [30]:
Variables_Entrenamiento=df[Variables_Entrenamiento_Posibles].select_dtypes(exclude="object").columns

In [31]:
for Variables_Arbol in Variables_Arboles:
    df_arbol=df[list(Variables_Entrenamiento)+[Variables_Arbol]]
    df_arbol_train=df_arbol[~df_arbol[Variables_Arbol].isnull()]
    df_arbol_test=df_arbol[df_arbol[Variables_Arbol].isnull()]
    if df_arbol[Variables_Arbol].dtypes not in ['float32','float64']:
        Modelo_C=DecisionTreeClassifier()
        print("no es flotante")
        if df_arbol[Variables_Arbol].dtypes=='O':
            print("La variable a predecir es tipo objeto")
            Encodar=LabelEncoder()
            df_arbol_train[Variables_Arbol]=Encodar.fit_transform(df_arbol_train[Variables_Arbol])
            print("termina el encoder")
        Modelo_C.fit(df_arbol_train[list(Variables_Entrenamiento)],df_arbol_train[Variables_Arbol])
        df_arbol_test[Variables_Arbol]=Modelo_C.predict(df_arbol_test[list(Variables_Entrenamiento)])
        if df_arbol[Variables_Arbol].dtypes=='O':
            df_arbol_train[Variables_Arbol]=Encodar.inverse_transform(df_arbol_train[Variables_Arbol])
            df_arbol_test[Variables_Arbol]=Encodar.inverse_transform(df_arbol_test[Variables_Arbol])
        df_conjunto=pd.concat([df_arbol_train,df_arbol_test]).sort_index()
        df[Variables_Arbol]=df_conjunto[Variables_Arbol]

    else:
        Modelo_R=DecisionTreeRegressor()
        print("no es flotante")
        if df_arbol_train[Variables_Arbol].dtypes=='O':
            print("Es tipo objeto")
            for columna in df_arbol_train.columns:
                if df_arbol_train[columna].dtypes=='O':
                    Encodar=LabelEncoder()
                    df_arbol_train[columna]=Encodar.fit_transform(df_arbol_train[columna])
                print("termina el encoder")
        Modelo_R.fit(df_arbol_train[list(Variables_Entrenamiento)],df_arbol_train[Variables_Arbol])
        df_arbol_test[Variables_Arbol]=Modelo_R.predict(df_arbol_test[list(Variables_Entrenamiento)])
        df_conjunto=pd.concat([df_arbol_train,df_arbol_test]).sort_index()
        df[Variables_Arbol]=df_conjunto[Variables_Arbol]

no es flotante
La variable a predecir es tipo objeto


C:\Users\Daniel Montes\AppData\Local\Temp\ipykernel_16952\1168305884.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_arbol_train[Variables_Arbol]=Encodar.fit_transform(df_arbol_train[Variables_Arbol])


termina el encoder


C:\Users\Daniel Montes\AppData\Local\Temp\ipykernel_16952\1168305884.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_arbol_test[Variables_Arbol]=Modelo_C.predict(df_arbol_test[list(Variables_Entrenamiento)])
C:\Users\Daniel Montes\AppData\Local\Temp\ipykernel_16952\1168305884.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_arbol_train[Variables_Arbol]=Encodar.inverse_transform(df_arbol_train[Variables_Arbol])
C:\Users\Daniel Montes\AppData\Local\Temp\ipykernel_16952\1168305884.py:17: Setting

no es flotante


C:\Users\Daniel Montes\AppData\Local\Temp\ipykernel_16952\1168305884.py:32: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_arbol_test[Variables_Arbol]=Modelo_R.predict(df_arbol_test[list(Variables_Entrenamiento)])


no es flotante
La variable a predecir es tipo objeto
termina el encoder


C:\Users\Daniel Montes\AppData\Local\Temp\ipykernel_16952\1168305884.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_arbol_train[Variables_Arbol]=Encodar.fit_transform(df_arbol_train[Variables_Arbol])
C:\Users\Daniel Montes\AppData\Local\Temp\ipykernel_16952\1168305884.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_arbol_test[Variables_Arbol]=Modelo_C.predict(df_arbol_test[list(Variables_Entrenamiento)])
C:\Users\Daniel Montes\AppData\Local\Temp\ipykernel_16952\1168305884.py:16: SettingWith

no es flotante


C:\Users\Daniel Montes\AppData\Local\Temp\ipykernel_16952\1168305884.py:32: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_arbol_test[Variables_Arbol]=Modelo_R.predict(df_arbol_test[list(Variables_Entrenamiento)])


no es flotante


C:\Users\Daniel Montes\AppData\Local\Temp\ipykernel_16952\1168305884.py:32: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_arbol_test[Variables_Arbol]=Modelo_R.predict(df_arbol_test[list(Variables_Entrenamiento)])


no es flotante


C:\Users\Daniel Montes\AppData\Local\Temp\ipykernel_16952\1168305884.py:32: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_arbol_test[Variables_Arbol]=Modelo_R.predict(df_arbol_test[list(Variables_Entrenamiento)])


no es flotante


C:\Users\Daniel Montes\AppData\Local\Temp\ipykernel_16952\1168305884.py:32: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_arbol_test[Variables_Arbol]=Modelo_R.predict(df_arbol_test[list(Variables_Entrenamiento)])


no es flotante


C:\Users\Daniel Montes\AppData\Local\Temp\ipykernel_16952\1168305884.py:32: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_arbol_test[Variables_Arbol]=Modelo_R.predict(df_arbol_test[list(Variables_Entrenamiento)])


no es flotante


C:\Users\Daniel Montes\AppData\Local\Temp\ipykernel_16952\1168305884.py:32: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_arbol_test[Variables_Arbol]=Modelo_R.predict(df_arbol_test[list(Variables_Entrenamiento)])


In [32]:
## Nos quedamos con la variable mayor cero para que sean nuestras variables depependientes
Var_Aux=pd.DataFrame(df.isnull().sum().sort_values(ascending=False)).reset_index().rename(columns={"index":"Variables",
                                                                                                           0:"Recuento"})
Var_Dependientes=Var_Aux[Var_Aux["Recuento"]>0]["Variables"].values
Var_Independientes=Var_Aux[Var_Aux["Recuento"]==0]["Variables"].values

In [33]:
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

In [53]:
Variables_KMEANS=Var_Dependientes[0]
df_kmeans=df.copy()
df_conjunto=pd.DataFrame()

for var_inde in Var_Independientes:
    if df_kmeans[var_inde].dtypes=='O':
        print("La variable a predecir es tipo objeto")
        Encodar=LabelEncoder()
        df_kmeans[var_inde]=Encodar.fit_transform(df_kmeans[var_inde])
        print("termina el encoder")
# COgemos una muestra para no tardar mucho
df_kmeans1=df_kmeans[Var_Independientes].sample(400,random_state=42)

# for Variable_KMEAN in [Variables_KMEANS]:
silueta_media=[]
for num_cluster in range(2,5):
    cluster=KMeans(n_clusters=num_cluster)
    cluster.fit(df_kmeans1)
    cluster_label=cluster.labels_
    silueta_media.append(silhouette_score(df_kmeans1,cluster_label))
Mejor_KMEANS=np.argmax(silueta_media)+2
##Una vez seleccionado
df_kmeans_Entreno=df_kmeans[Var_Independientes]
cluster=KMeans(n_clusters=Mejor_KMEANS)
cluster.fit(df_kmeans_Entreno)
df_kmeans_Entreno["Cluster"]=cluster.predict(df_kmeans_Entreno)
for Variable_KMEAN in [Variables_KMEANS]:
    df_kmeans_Entreno[Variable_KMEAN]=df[Variable_KMEAN]
    for cluster_num in df_kmeans_Entreno["Cluster"].unique():
        print(f"{cluster_num}")
        clusterizado=df_kmeans_Entreno[df_kmeans_Entreno["Cluster"]==cluster_num].drop(columns="Cluster")
        clusterizado_train=clusterizado[~clusterizado[Variable_KMEAN].isnull()]
        clusterizado_test=clusterizado[clusterizado[Variable_KMEAN].isnull()]
        if df[Variable_KMEAN].dtypes not in ['float32','float64']:
            print("Clasificacion")
            Modelo_C=DecisionTreeClassifier()
            Modelo_C.fit(clusterizado_train[Var_Independientes],clusterizado_train[Variable_KMEAN])
            clusterizado_test[Variable_KMEAN]=Modelo_C.predict(clusterizado_test[Var_Independientes])
        else:
            print("Regresion")
            Modelo_R=DecisionTreeRegressor()
            Modelo_R.fit(clusterizado_train[Var_Independientes],clusterizado_train[Variable_KMEAN])
            clusterizado_test[Variable_KMEAN]=Modelo_R.predict(clusterizado_test[Var_Independientes])
    
        df_conjunto_Aux=pd.concat([clusterizado_train,clusterizado_test])
    df_conjunto=pd.concat([df_conjunto,df_conjunto_Aux]).sort_index()
    df[Variable_KMEAN]=df_conjunto[Variable_KMEAN]        

La variable a predecir es tipo objeto
termina el encoder
La variable a predecir es tipo objeto
termina el encoder
La variable a predecir es tipo objeto
termina el encoder
La variable a predecir es tipo objeto
termina el encoder
La variable a predecir es tipo objeto
termina el encoder
La variable a predecir es tipo objeto
termina el encoder
La variable a predecir es tipo objeto
termina el encoder
La variable a predecir es tipo objeto
termina el encoder
La variable a predecir es tipo objeto
termina el encoder
La variable a predecir es tipo objeto
termina el encoder
La variable a predecir es tipo objeto
termina el encoder
La variable a predecir es tipo objeto
termina el encoder


C:\Users\Daniel Montes\AppData\Local\Temp\ipykernel_16952\1135729576.py:26: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_kmeans_Entreno["Cluster"]=cluster.predict(df_kmeans_Entreno)
C:\Users\Daniel Montes\AppData\Local\Temp\ipykernel_16952\1135729576.py:28: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_kmeans_Entreno[Variable_KMEAN]=df[Variable_KMEAN]


0
Regresion


C:\Users\Daniel Montes\AppData\Local\Temp\ipykernel_16952\1135729576.py:43: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clusterizado_test[Variable_KMEAN]=Modelo_R.predict(clusterizado_test[Var_Independientes])


1
Regresion


ValueError: Found array with 0 sample(s) (shape=(0, 75)) while a minimum of 1 is required.

In [59]:
df_kmeans_Entreno[df_kmeans_Entreno["Cluster"]==1][Variables_KMEANS].value_counts(dropna=False)

NaN    96613
Name: FLOORSMAX_AVG, dtype: int64

In [60]:
Modelo_R=DecisionTreeRegressor()
Modelo_R.fit(df_kmeans_Entreno[Var_Independientes],df_kmeans_Entreno["Cluster"])

DecisionTreeRegressor()

In [63]:
Diccionario={}
for x,y in zip(Modelo_R.feature_names_in_,Modelo_R.feature_importances_):
    if y>0:
        Diccionario[x]=y

In [67]:
sorted(Diccionario.items(),key=lambda x:x[1])

[('REG_REGION_NOT_LIVE_REGION', 2.9703761793158743e-06),
 ('AMT_REQ_CREDIT_BUREAU_YEAR', 7.5461012969676116e-06),
 ('NAME_INCOME_TYPE', 7.847956253792069e-06),
 ('FLAG_WORK_PHONE', 1.194108337102567e-05),
 ('NAME_EDUCATION_TYPE', 1.2576835494946019e-05),
 ('REGION_RATING_CLIENT_W_CITY', 1.2576835494946019e-05),
 ('TARGET', 1.3931263932863267e-05),
 ('DEF_60_CNT_SOCIAL_CIRCLE', 1.437352627993831e-05),
 ('AMT_REQ_CREDIT_BUREAU_MON', 1.4373635693346744e-05),
 ('OBS_30_CNT_SOCIAL_CIRCLE', 1.8110643112722267e-05),
 ('ORGANIZATION_TYPE', 1.8293578901739665e-05),
 ('AMT_REQ_CREDIT_BUREAU_WEEK', 2.1893009935646773e-05),
 ('WEEKDAY_APPR_PROCESS_START', 2.4147524150296348e-05),
 ('NAME_HOUSING_TYPE', 2.704019631413394e-05),
 ('HOUR_APPR_PROCESS_START', 3.169362544726397e-05),
 ('OWN_CAR_AGE', 3.647282293534345e-05),
 ('CNT_CHILDREN', 3.760577753778074e-05),
 ('REGION_POPULATION_RELATIVE', 3.848511661453481e-05),
 ('EXT_SOURCE_2', 4.0451465485224506e-05),
 ('DAYS_LAST_PHONE_CHANGE', 4.23742611291